linera-core/src/worker.rs

```diff
@@ -455,27 +456,57 @@ where
         self.key_pair.as_ref().map(Arc::as_ref)
     }
 
+    /// Creates an `UpdateRecipient` request that informs the `recipient` about new
+    /// cross-chain messages from `sender`.
     async fn create_cross_chain_request(
         &self,
         confirmed_log: &LogView<StorageClient::Context, CryptoHash>,
         height_map: Vec<(Medium, Vec<BlockHeight>)>,
         sender: ChainId,
         recipient: ChainId,
     ) -> Result<CrossChainRequest, WorkerError> {
+        // Load all the certificates we will need, regardless of the medium.
         let heights =
             BTreeSet::from_iter(height_map.iter().flat_map(|(_, heights)| heights).copied());
-        let mut heights_usize = Vec::new();
-        for height in heights {
-            heights_usize.push(height.try_into()?);
-        }
-        let values = confirmed_log.multi_get(heights_usize).await?;
-        let keys = values.into_iter().flatten();
-        let certificates = self.storage.read_certificates(keys).await?;
+        let heights_usize = heights
+            .iter()
+            .copied()
+            .map(usize::try_from)
+            .collect::<Result<Vec<_>, _>>()?;
```

Comment:
The for-loop took 4 lines whereas this is 5 lines but ok :) Also, do we know that the generated code is the same?